tests/integration/integration_suite_test.go

```diff
@@ -1,13 +1,326 @@
+			namespace := initializeCluster(t)
+			createMedusaSecretAndInstallDeps(t, namespace, backend)
+			deployClusterForMedusa(t, namespace, backend)
+			medusaSuccess := t.Run("Test backup and restore", func(t *testing.T) {
+				testMedusa(t, namespace, backend, backupName)
+			})
+			if medusaSuccess {
+				cleanupCluster(t, namespace)
+			}
+		})
+	}
+}
+
+func testMedusa(t *testing.T, namespace, backend, backupName string) {
+	log.Println(Step("Testing Medusa..."))
+	log.Println("Creating test keyspace and table")
+	CreateCassandraTable(t, namespace, medusaTestTable, medusaTestKeyspace)
+
+	loadRowsAndCheckCount(t, namespace, 10, 10)
+
+	log.Println(Info("Backing up Cassandra"))
+	PerformBackup(t, namespace, backupName)
+
+	loadRowsAndCheckCount(t, namespace, 10, 20)
+
+	log.Println(Info("Restoring backup and checking row count"))
+	RestoreBackup(t, namespace, backupName)
+	CheckRowCountInTable(t, 10, namespace, medusaTestTable, medusaTestKeyspace)
+}
+
+func deployClusterForMedusa(t *testing.T, namespace, backend string) {
+	log.Println(Info(fmt.Sprintf("Deploying K8ssandra with Medusa using %s", backend)))
+	valuesFile := fmt.Sprintf("one_node_cluster_with_medusa_%s.yaml", strings.ToLower(backend))
+	DeployClusterWithValues(t, namespace, strings.ToLower(backend), valuesFile)
+	CheckClusterExpectedResources(t, namespace)
+}
+
+func loadRowsAndCheckCount(t *testing.T, namespace string, rowsToLoad, rowsExpected int) {
+	log.Println(Info(fmt.Sprintf("Loading %d rows and checking we have %d after", rowsToLoad, rowsExpected)))
+	LoadRowsInTable(t, rowsToLoad, namespace, medusaTestTable, medusaTestKeyspace)
+	CheckRowCountInTable(t, rowsExpected, namespace, medusaTestTable, medusaTestKeyspace)
+}
+
+func createMedusaSecretAndInstallDeps(t *testing.T, namespace, backend string) {
+	log.Println(Info("Creating medusa secret to access the backend"))
+	if backend == "Minio" {
+		DeployMinioAndCreateBucket(t, "k8ssandra-medusa")
+		CreateMedusaSecretWithFile(t, namespace, "secret/medusa_minio_secret.yaml")
+	} else {
+		// S3
+		CreateMedusaSecretWithFile(t, namespace, "~/medusa_secret.yaml")
+	}
+}
+
+// Monitoring scenario:
+// - Install Traefik
+// - Create a namespace
+// - Register a cluster with three Cassandra nodes and one Stargate node
+// - Check that Prometheus is reachable through its REST API
+// - Check the number of active Prometheus targets
+// - Check that Grafana is reachable through http
+// - Terminate the namespace and delete the cluster
+func TestMonitoringDeploymentScenario(t *testing.T) {
+	// only runs when the -short flag is used
+	if !testing.Short() {
+		t.Skip("Skipped in non-short mode")
+	}
+	namespace := initializeCluster(t)
+	deployClusterForMonitoring(t, namespace)
+	prometheusSuccess := t.Run("Test Prometheus", func(t *testing.T) {
+		testPrometheus(t, namespace)
+	})
+
+	grafanaSuccess := t.Run("Test Grafana", func(t *testing.T) {
+		testGrafana(t, namespace)
+	})
+
+	if prometheusSuccess && grafanaSuccess {
```

Comment:
So if a test fails, the cluster is left to a disarray and user must clean it up?

Reply:
My opinion, based on early integration tests constructed, is to consider a full `cluster cleanup` as `teardown` step and `cluster setup` as `setup` step.

If the `scope of the test` is to verify a specific delete/removal use case, then prior to full teardown step, we would verify that there wasn't an error and the resource is properly cleaned up.

Reply:
yes, and that's necessary for investigation purposes (in and out of the CI context, but specifically for CI we need to keep the cluster running). When running the integration tests with `make kind-integ-test`, the kind cluster will get deleted before being recreated.

Reply:
@jeffbanks, I see a few issues there (which could be due to my Go skills):

- The `initializeCluster()` function (which is what would be in `setUp()`) generates a unique namespace name which is passed on to all the test functions and there were concerns raised previously with the use of global variables. Now if we can pass the generated value on from `TestMain()` to the individual tests, then that's a good thing to do as it would reduce code duplication.
- The `initializeCluster()` and `cleanupCluster()` functions use Terratest underneath which requires to use the `testing.T` object which doesn't seem to be available in the `TestMain()` function. It seems like the only thing we can get out of the individual tests is the return code.

Unless we have an easy solution to do this (which doesn't involve removing Terratest, which wouldn't be a small change) then I'd like to defer this as part of a future refactoring.

What do you think?

Reply:
👍 I agree, let's address this moving forward, I was able to use `Terratest`  with a combination of setup and teardowns by managing the scope of context and namespace.   I am good with iterating and refactoring some of this as we understand more of what we want.  IMO, this PR has become massive, with too many working parts.  :smile: